src/Arrays/MPIStateArrays.jl

```diff
@@ -258,91 +258,38 @@ end
 
 # Integral based metrics
 function LinearAlgebra.norm(Q::MPIStateArray; p::Real=2)
+  T = eltype(Q)
+  rQ = @view Q.Q[:, :, Q.realelems]
 
-  @assert p == 2
-
-  host_array = Array ∈ typeof(Q).parameters
-  h_Q = host_array ? Q : Array(Q)
-  Np = size(Q, 1)
-
-  if isempty(Q.weights)
-    locnorm2 = knl_norm2(Val(Np), h_Q, Q.realelems)
+  if isfinite(p)
+    E = @~ abs.(rQ).^p
+    op, mpiop, init = +, MPI.SUM, zero(T)
   else
-    locnorm2 = knl_L2norm(Val(Np), h_Q, Q.weights, Q.realelems)
+    E = @~ abs.(rQ)
+    op, mpiop, init = max, MPI.MAX, typemin(T)
   end
 
-  sqrt(MPI.Allreduce([locnorm2], MPI.SUM, Q.mpicomm)[1])
-end
-
-function knl_norm2(::Val{Np}, Q, elems) where {Np}
-  DFloat = eltype(Q)
-  (_, nstate, nelem) = size(Q)
-
-  nrm = zero(DFloat)
-
-  @inbounds for e = elems, q = 1:nstate, i = 1:Np
-    nrm += Q[i, q, e]^2
-  end
-
-  nrm
-end
-
-function knl_L2norm(::Val{Np}, Q, weights, elems) where {Np}
-  DFloat = eltype(Q)
-  (_, nstate, nelem) = size(Q)
-
-  nrm = zero(DFloat)
-
-  @inbounds for e = elems, q = 1:nstate, i = 1:Np
-    nrm += weights[i, e] * Q[i, q, e]^2
+  if ~isempty(Q.weights)
+    w = @view Q.weights[:, :, Q.realelems]
+    E = @~ E .* w
   end
 
-  nrm
+  locnorm = mapreduce(identity, op, E, init=init)
+  sqrt(MPI.Allreduce([locnorm], mpiop, Q.mpicomm)[1])
```

Comment:
This sqrt is only correct for p == 2, if you want this function to work for general p you need x ^ (1 / p) and handle the case of p == Inf.

Reply:
Good catch, thanks!